experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/editableTree.events.spec.ts

```diff
@@ -0,0 +1,414 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { strict as assert } from "assert";
+
+import { MockFluidDataStoreRuntime } from "@fluidframework/test-runtime-utils";
+
+import { FieldKinds, SchemaBuilder } from "../../../feature-libraries";
+import { AllowedUpdateType } from "../../../core";
+import { TypedTreeFactory } from "../../../typed-tree";
+import { ForestType } from "../../../shared-tree";
+import { typeboxValidator } from "../../../external-utilities";
+import { leaf } from "../../..";
+
+describe("beforeChange/afterChange events", () => {
+	const builder = new SchemaBuilder({
+		scope: "beforeChange/afterChange events",
+		libraries: [leaf.library],
+	});
+	const myInnerNodeSchema = builder.struct("myInnerNode", {
+		myInnerString: SchemaBuilder.required(leaf.string),
+	});
+	const myNodeSchema = builder.structRecursive("myNode", {
+		child: SchemaBuilder.required(myInnerNodeSchema),
+		myString: SchemaBuilder.required(leaf.string),
+		myOptionalNumber: SchemaBuilder.optional(leaf.number),
+	});
+	const schema = builder.toDocumentSchema(SchemaBuilder.field(FieldKinds.required, myNodeSchema));
+
+	it("fire the expected number of times", () => {
+		// TODO: once assignment to properties is implemented in EditableTree2, update this test to apply changes like
+		//   root.myString = "new string";
+		// instead of
+		//   root.boxedMyString.content = "new string";
+
+		const factory = new TypedTreeFactory({
+			jsonValidator: typeboxValidator,
+			forest: ForestType.Reference,
+			allowedSchemaModifications: AllowedUpdateType.None,
+			initialTree: {
+				myString: "initial string",
+				myOptionalNumber: 3,
+				child: { myInnerString: "initial string in child" },
+			},
+			schema,
+			subtype: "test",
+		});
+
+		const root = factory.create(new MockFluidDataStoreRuntime(), "the tree").root.content;
```

Comment:
I've been trying (unsuccessfully) to extract this to a function that can be reused by all tests (each providing an initial tree appropriate for what it will test). In particular can't figure out how to type the parameter for the initial tree. Any pointers welcome :)

Reply:
With the new pattern of the initial tree being passed to a `.schematize()` call on the object returned by the factory I was able to achieve most of this and it feels good enough.